newsroom/push.py

```diff
@@ -87,17 +121,136 @@ def publish_item(doc):
-    app.cache.delete(HOME_ITEMS_CACHE_KEY)
-    return flask.jsonify({})
+def publish_event(event):
+    orig = app.data.find_one('agenda', req=None, _id=event['guid'])
+    service = superdesk.get_resource_service('agenda')
+
+    if not orig:
+        # new event
+        agenda = {}
+        agenda.setdefault('_id', event['guid'])
+        agenda['event_id'] = event['guid']
+        agenda['name'] = event.get('name')
+        agenda['slugline'] = event.get('slugline')
+        agenda['definition_long'] = event.get('definition_long')
+        agenda['calendars'] = event.get('calendars')
+        agenda['location'] = event.get('location')
+        agenda['event'] = event
+        now = utcnow()
+        parse_dates(event)
+        agenda.setdefault('firstcreated', now)
+        agenda.setdefault('versioncreated', now)
+        agenda.setdefault(app.config['VERSION'], 1)
+
+        agenda['dates'] = get_event_dates(event)
+
+        _id = service.create([agenda])[0]
+
+    else:
+        # replace the original document
+        if event.get('state') in [WORKFLOW_STATE.CANCELLED, WORKFLOW_STATE.KILLED] or \
+                event.get('pubstatus') == 'cancelled':
+            # it has been cancelled so don't need to change the dates
+            updates = {'event': event}
+            _id = service.patch(event['guid'], updates)
+        elif event.get('state') in [WORKFLOW_STATE.RESCHEDULED, WORKFLOW_STATE.POSTPONED]:
+            # schedule is changed, recalculate the dates, planning id and coverages from dates will be removed
+            updates = {'event': event}
+            updates['dates'] = get_event_dates(event)
+            _id = service.patch(event['guid'], updates)
+
+    return _id
+
+
+def get_event_dates(event):
+    dates = []
+
+    event['dates']['start'] = datetime.strptime(event['dates']['start'], '%Y-%m-%dT%H:%M:%S+0000')
+    event['dates']['end'] = datetime.strptime(event['dates']['end'], '%Y-%m-%dT%H:%M:%S+0000')
+
+    if event['dates'].get('recurring_rule', None):
+        # recurring event
+        recurring_events = generate_recurring_events(event)
+        for recurring_event in recurring_events:
+            dates.append({
+                'start': recurring_event['dates']['start'],
+                'end': recurring_event['dates']['end'],
+                'state': recurring_event['state'],
+            })
+    else:
+        # non recurring event
+        dates.append({
+            'start': event['dates']['start'],
+            'end': event['dates']['end'],
+            'state': event['state'],
+        })
+    return dates
+
+
+def publish_planning(planning):
+    service = superdesk.get_resource_service('agenda')
+
+    # update dates
+    planning['planning_date'] = datetime.strptime(planning['planning_date'], '%Y-%m-%dT%H:%M:%S+0000')
+
+    if planning.get('event_item'):
+        # this is a planning for an event item
+        agenda = app.data.find_one('agenda', req=None, _id=planning['event_item'])
```

Comment:
same here probably, search by `event_id` would make more sense imo

Reply:
Like the one above, I'm planning to add a comment here like `if there's an event then _id field will have the same value as event_id` ...